examples/setixset.dx

```diff
@@ -0,0 +1,190 @@
+    -- functon to operate in parallel, but it would be
+    -- wasteful in the case where there is an early difference.
+    iter \i.
+      case i == min nx ny of
+        True -> Done $ compareLengths nx ny
+        False ->
+          xi = xs.(unsafeFromOrdinal _ i)
+          yi = ys.(unsafeFromOrdinal _ i)
+          case compareElements xi yi of
+            True -> Done True
+            False -> case xi == yi of
+              True -> Continue
+              False -> Done False
+
+instance {n} [Ord n] Ord (List n)
+  (>) = lexicalOrder (>) (>)
+  (<) = lexicalOrder (<) (<)
+
+:p "aaa" < "bbb"
+> True
+
+:p "aa" < "bbb"
+> True
+
+:p "a" < "aa"
+> True
+
+:p "aaa" > "bbb"
+> False
+
+:p "aa" > "bbb"
+> False
+
+:p "a" > "aa"
+> False
+
+:p "a" < "aa"
+> True
+
+:p ("": List Word8) > ("": List Word8)
+> False
+
+:p ("": List Word8) < ("": List Word8)
+> False
+
+:p "a" > "a"
+> False
+
+:p "a" < "a"
+> False
+
+:p "Thomas" < "Thompson"
+> True
+
+:p "Thomas" > "Thompson"
+> False
+
+
+
+:p sort [1]
+
+:p sort [3, 2, 1]
+
+:p sort ["A"]
+
+:p sort ["A", "B", "C"]
+
+:p sort ["Alice", "Bob", "Charlie"]
+
+:p sort ["Charlie", "Alice", "Bob", "Aaron"]
+
+----------------------
+
+
+'#### Misc Ix Set instances
+
+
+instance Ix Bool
+  getSize = \(). 2
+  ordinal = \b. case b of
+    False -> 0
+    True -> 1
+  unsafeFromOrdinal = \k. k == 1
+
+:p for i:Bool.
+  i
+
+-----------------
+
+
+'### An index set for List
+
+data ListIx a l:(List a) =
+  MkListIx Int 
+
+instance {a list} Ix (ListIx a list)
+  getSize = \(). listLength list
+  ordinal = \(MkListIx i). i
+  unsafeFromOrdinal = \k. MkListIx k
+
+names = AsList _ ["Alice", "Bob", "Charlie"]
+enumerateNames : (ListIx _ names)=>Int =
+  for i. ordinal i  -- expensive round trip!?
+
+:p enumerateNames
+:p size (ListIx _ names)
+
+:p for i:(ListIx _ names). i
+
+
+
+
+
+
+
+'### An index set for Sets
+
+'#### Sets
+
+data Set a [Ord a] =
+  -- Guaranteed to be in sorted order,
+  -- as long as no one else uses this constructor.
+  -- Instead use the "toSet" function below.
+  UnsafeAsSet n:Int elements:(Fin n => a)
+
+def toSet {n a} [Ord a] (xs:n=>a) : Set a =
+  n' = size n
+  sorted_xs = sort xs
+  UnsafeAsSet _ $ unsafeCastTable (Fin n') sorted_xs
+
+def setLength {a} ((UnsafeAsSet n _):Set a) : Int = n
+
+instance {a} [Eq a] Eq (Set a)
+  (==) = \(UnsafeAsSet nx xs) (UnsafeAsSet ny ys).
+    if nx /= ny
+      then False
+      else all for i:(Fin nx).
+        xs.i == ys.(unsafeFromOrdinal _ (ordinal i))
+
+
+'##### Tests
+
+-- check order invariance.
+:p (toSet ["Bob", "Alice", "Charlie"]) == (toSet ["Charlie", "Bob", "Alice"])
+
+
+'#### Index set for sets of strings
+
+--data SetIx a l:(Set a) [Ord a] =
+-- need a bugfix before we can be polymorphic in type
+
+data StringSetIx l:(Set String) =
+  MkSetIx Int
+
+instance {set} Ix (StringSetIx set)
+  getSize = \(). setLength set
+  ordinal = \(MkSetIx i). i
+  unsafeFromOrdinal = \k. MkSetIx k
```

Comment:
I think that we we additionally need form those list/set index sets is a way to convert an ordinal into an element that it corresponds to. You would write something along the lines of:
```
interface AssocIx n  -- index sets where indices have data associated with them
  IxValueType : Type
  ixValue : n -> IxValueType n
```
Of course for now we can hardcode the associated type, because I'm not sure if they work properly? But might be worth trying.

Reply:
Agreed.  Dougal had suggested that we need the inverse function, a lookup that gives the index given the data.  So maybe more like:

```
interface AssocIx n  -- index sets where indices have data associated with them
  IxValueType : Type
  ixValue : n -> IxValueType n
  lookupIx : IxValueType n -> n
```

But in any case, this doesn't currently work, as you guessed.  But in the meantime, I'll code up the relevant functions for `Set`.

Reply:
Okay I've added these functions, `setIxToString` and `stringToSetIx`.